ranger-discovery-bundle/src/main/java/io/appform/ranger/discovery/bundle/id/DistributedIdGenerator.java

```diff
@@ -0,0 +1,288 @@
+        partitionCount = partitionSize;
+        partitionResolver = partitionResolverSupplier;
+        idFormatter = idFormatterInstance;
+    }
+
+    public synchronized void registerGlobalConstraints(final KeyValidationConstraint... constraints) {
+        registerGlobalConstraints(ImmutableList.copyOf(constraints));
+    }
+
+    public synchronized void registerGlobalConstraints(final List<KeyValidationConstraint> constraints) {
+        Preconditions.checkArgument(null != constraints && !constraints.isEmpty());
+        GLOBAL_CONSTRAINTS.addAll(constraints);
+    }
+
+    public synchronized void registerDomainSpecificConstraints(
+            final String domain,
+            final KeyValidationConstraint... validationConstraints) {
+        registerDomainSpecificConstraints(domain, ImmutableList.copyOf(validationConstraints));
+    }
+
+    public synchronized void registerDomainSpecificConstraints(
+            final String domain,
+            final List<KeyValidationConstraint> validationConstraints) {
+        Preconditions.checkArgument(null != validationConstraints && !validationConstraints.isEmpty());
+        DOMAIN_SPECIFIC_CONSTRAINTS.computeIfAbsent(domain, key -> new ArrayList<>())
+                .addAll(validationConstraints);
+    }
+
+    /**
+     * Generate id with given prefix
+     *
+     * @param prefix String prefix for ID to be generated
+     * @return Generated Id
+     */
+    public Id generate(final String prefix) {
+        val targetPartitionId = getTargetPartitionId();
+        return generateForPartition(prefix, targetPartitionId);
+    }
+
+    public Id generateForPartition(final String prefix, final int targetPartitionId) {
+        val prefixIdMap = dataStore.computeIfAbsent(prefix, k -> new ConcurrentHashMap<>());
+        val currentTimestamp = new DateTime();
+        val timeKey = currentTimestamp.getMillis() / 1000;
+        val idCounter = generateForAllPartitions(
+                prefixIdMap.computeIfAbsent(timeKey, key -> new PartitionIdTracker(partitionCount)),
+                prefix,
+                currentTimestamp,
+                targetPartitionId);
+        val id = String.format("%s%s", prefix, idFormatter.format(currentTimestamp, NODE_ID, idCounter));
+        return Id.builder()
+                .id(id)
+                .exponent(idCounter)
+                .generatedDate(currentTimestamp.toDate())
+                .node(NODE_ID)
+                .build();
+    }
+
+    private int generateForAllPartitions(final PartitionIdTracker partitionIdTracker,
+                                         final String prefix,
+                                         final DateTime timestamp,
+                                         final int targetPartitionId) {
+        val idPool = partitionIdTracker.getPartition(targetPartitionId);
+        int idIdx = idPool.getPointer().getAndIncrement();
+//            ToDo: Add Retry Limit
+        while (idPool.getIds().size() <= idIdx) {
```

Comment:
Let's add a retry limit which is explicitly passed by client, since partitionResolver is passed by client.